src/algebra/order/interval.lean

```diff
@@ -0,0 +1,336 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import algebra.order.group.basic
+import data.option.n_ary
+import data.set.pointwise.basic
+import order.interval
+
+/-!
+# Interval arithmetic
+
+This file defines arithmetic operations on intervals and prove their correctness. Note that this is
+full precision operations. We do not yet have float operations.
+-/
+
+open function set
+open_locale pointwise
+
+variables {α : Type*}
+
+/-! ### One/zero -/
+
+section one
+section preorder
+variables [preorder α] [has_one α]
+
+@[to_additive] instance : has_one (nonempty_interval α) := ⟨nonempty_interval.pure 1⟩
+@[to_additive] instance : has_one (interval α) := ⟨interval.pure 1⟩
+
+namespace nonempty_interval
+
+@[simp, to_additive to_prod_zero] lemma to_prod_one : (1 : nonempty_interval α).to_prod = 1 := rfl
+@[to_additive] lemma fst_one : (1 : nonempty_interval α).fst = 1 := rfl
+@[to_additive] lemma snd_one : (1 : nonempty_interval α).snd = 1 := rfl
+@[simp, norm_cast, to_additive] lemma coe_one : ((1 : nonempty_interval α) : interval α) = 1 := rfl
+@[simp, to_additive] lemma pure_one : pure (1 : α) = 1 := rfl
+
+end nonempty_interval
+
+namespace interval
+
+@[simp, to_additive] lemma pure_one : pure (1 : α) = 1 := rfl
+@[simp, to_additive] lemma one_ne_bot : (1 : interval α) ≠ ⊥ := pure_ne_bot
+@[simp, to_additive] lemma bot_ne_one : (⊥ : interval α) ≠ 1 := bot_ne_pure
+
+end interval
+end preorder
+
+section partial_order
+variables [partial_order α] [has_one α]
+
+@[simp, to_additive]
+lemma nonempty_interval.coe_set_one : ((1 : nonempty_interval α) : set α) = 1 := Icc_self _
+@[simp, to_additive] lemma interval.coe_one : ((1 : interval α) : set α) = 1 := Icc_self _
+
+end partial_order
+end one
+
+/-!
+### Addition/multiplication
+
+Note that this multiplication does not apply to `ℚ` or `ℝ`.
+-/
+
+section mul
+variables [preorder α] [has_mul α] [covariant_class α α (*) (≤)]
+  [covariant_class α α (swap (*)) (≤)]
+
+@[to_additive] instance : has_mul (nonempty_interval α) :=
+⟨λ s t, ⟨s.to_prod * t.to_prod, mul_le_mul' s.fst_le_snd t.fst_le_snd⟩⟩
+
+@[to_additive] instance : has_mul (interval α) := ⟨option.map₂ (*)⟩
+
+namespace nonempty_interval
+variables (s t : nonempty_interval α) (a b : α)
+
+@[simp, to_additive to_prod_add] lemma to_prod_mul : (s * t).to_prod = s.to_prod * t.to_prod := rfl
+@[to_additive] lemma fst_mul : (s * t).fst = s.fst * t.fst := rfl
+@[to_additive] lemma snd_mul : (s * t).snd = s.snd * t.snd := rfl
+@[simp, to_additive] lemma coe_mul : (↑(s * t) : interval α) = s * t := rfl
+@[simp, to_additive] lemma pure_mul_pure : pure a * pure b = pure (a * b) := rfl
+
+end nonempty_interval
+
+namespace interval
+variables (s t : interval α)
+
+@[simp, to_additive] lemma bot_mul : ⊥ * t = ⊥ := rfl
+@[simp, to_additive] lemma mul_bot : s * ⊥ = ⊥ := option.map₂_none_right _ _
+
+end interval
+end mul
+
+/-! ### Powers -/
+
+-- TODO: Why is `to_additive` failing here?
```

Comment:
I tried troubleshooting but didn't find any meaningful progress. I think the changed order of parameters means you just have to go for manually defining this.

Reply:
```suggestion
-- TODO: if `to_additive` gets improved sufficiently, derive this from `has_pow`
```